### PR TITLE
[Draft] os/bluestore/bluefs: Add ability for direct write mode

### PR DESCRIPTION
**Pull request by aclamk**
Makes possible to use direct write mode and buffered read mode.
Option bluefs_force_direct_write=true uses direct write path, regardless of
value of bluefs_buffered_io. Bluefs_buffered_io is still in effect for signaling RocksDB.

Signed-off-by: Adam Kupczyk <akupczyk@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by aclamk** [comment]
Performance comparision (RH only):
https://docs.google.com/spreadsheets/d/166WkxYQriq_wX7XqlGrfSzdtYHxh5Q622Ux5LoyfxMI/edit#gid=0

**Comment by aclamk** [comment]
Results on officinalis-8 testing, short.

Case 1: default
bluefs_buffered_io = true, bluefs_force_direct_write = false
run | write | read   
-----|-------|-------------
 0 | 12408 |	101644  |
 1 | 12655 |	99754   |
 2 | 12315 |	96148   |
 3 | 12108 |	94296   |
 4 | 11960 |	93107   |
 5 | 11815 |	95617   |
 6 | 11637 |	92906   |
 7 | 11310 |	91014   |
 lat | 8549 |	26129   |

Case 2:
bluefs_buffered_io = true, bluefs_force_direct_write = true
 run  | write |	read
---|-------|----------
 0 | 40275 |	102773
 1 | 44715 |	101931
 2 | 44246 |	94168
 3 | 45209 |	95782
 4 | 44512 |	90382
 5 | 45296 |	100049
 6 | 45470 |	92971
 7 | 44570 |	89986
 lat | 12412 |	25570

Case 3:
bluefs_buffered_io = false
run   | write |	read
---|------|----
0 | 29967 |	102185
 1 | 28905 |	101590
 2 | 28885 |	93862
 3 | 28812 |	92262
 4 | 28257 |	84212
 5 | 27213 |	92115
 6 | 27286 |	90649
 7 | 28717 |	82574
 lat | 11825 |	25106

File: src/os/bluestore/BlueFS.cc [hunk 1]
```diff
@@ -2582,6 +2602,11 @@ int BlueFS::_flush_and_sync_log(std::unique_lock<ceph::mutex>& l,
     if (!to_release[i].empty()) {
       /* OK, now we have the guarantee alloc[i] won't be null. */
       int r = 0;
+      if (forced_direct_writes == true) {
```
Comment (ifed01) [inline src/os/bluestore/BlueFS.cc:2605]: AFAIU this block is orthogonal to direct writes - this is primarily intended to drop cached data, i.e. one that has been read.
Shouldn't this be controlled by another setting then?

File: src/os/bluestore/BlueFS.cc [hunk 2]
```diff
@@ -815,6 +815,26 @@ int BlueFS::mount()
...
+      cct->_conf->bluefs_buffered_io == true) {
+    /* reads buffered, writes direct */
+    bool dontneed_check = true;
+    for (unsigned dev = BDEV_WAL; dev <= BDEV_SLOW; dev++) {
+      if (bdev[dev]) {
+	int r = bdev[dev]->dontneed(0, super.block_size);
+	if (r != 0) {
+	  dout(1) << __func__ << " bluefs_force_direct_write selected, but "
+		  << get_device_name(dev) << " cant support it" << dendl;
+	  dontneed_check = false;
```
Comment (ifed01) [inline src/os/bluestore/BlueFS.cc:828]: break at this point

**Comment by ifed01** [comment]
> Results on officinalis-8 testing, short. Values in IO/s.
> 
> Case 2:
> bluefs_buffered_io = true, bluefs_force_direct_write = true
> 
> run	write	read
> 0	40275	102773
...
> Case 3:
> bluefs_buffered_io = false
> 
> run	write	read
> 0	29967	102185
> 1	28905	101590

Wondering if you have any understanding why case 2 has got much faster writes?

**Comment by ifed01** [comment]
Looks interesting!

**Comment by ifed01** [comment]
One concern though - do you understand how consistent is the simultaneous usage of both buffered and direct I/O. Would buffered reads get proper data immediately after the relevant writing is completed if data to read is present in the cache?

As a side note I have some feeling that we might already have some issues of that sort when using from containers...

**Comment by satoru-takeuchi** [comment]
I'm afraid to introduce more complexity about direct I/O into bluefs I/O.
Buffered I/O and direct I/O shouldn't be used for the same devices/files
since it might cause data inconsistency.

https://man7.org/linux/man-pages/man2/open.2.html
> Applications should avoid mixing O_DIRECT and normal I/O to the
> same file, and especially to overlapping byte regions in the same file.

In addition, direct I/O is generally used to implement application's own cache
instead of page/buffer cache.

> Try to minimize cache effects of the I/O to and from this file. 
> In general this will degrade performance, but it is useful in special situations,
> such as when applications do their own caching.

So, I think it's better to improve bluefs's own cache mechanism.

**Comment by aclamk** [comment]
@satoru-takeuchi 
In principle, I agree that frivolous mixing of direct I/O and buffered I/O is a bad practice, and should be avoided.
The reason why I even exercise this mixing mode is that it is much faster to:
- aio_write()
- aio_wait()
- fdatasync()
then
- write()
- sync_file_range()
- fdatasync()
I tested this on 3 machines. This is the effect no caching can speed up.

BlueFS is special in the sense that it fully controls all path to data.
While current proposal is faulty (re-reading data block that has been appended to bluefs file might be corrupted), 
in general we could fully control to always purge cache region before reading.

**Comment by wjwithagen** [comment]
Hi @aclamk 

Care to share the code of the tests you did, and perhaps some of the results.
So I can try them on FreeBSD, and see if the magnitude of the results is in the same ballpark as on Linux....
Even the comparison of the 2 tests by themselves will show if the same holds on FreeBSD

**Comment by aclamk** [comment]
@wjwithagen 
I use https://github.com/aclamk/incerta-testing as test environment.
It uses strongly skewed pareto distribution for object access and io sizes, making some objects strongly preferable and making smaller IOs happen more often then larger ones.
Porting it will probably be irritating work, sorry.
Main points are:
- uses 4 same FIO jobs
- uses rados FIO engine
- sets "--touch_objects=0" to avert obfuscating Onode cache init
- sets IO sizes: "--bssplit=4k/16:8k/10:12k/9:16k/8:20k/7:24k/7:28k/6:32k/6:36k/5:40k/5:44k/4:48k/4:52k/4:56k/3:60k/3:64k/3"
- sets file_service_type=pareto:0.20:0

**Comment by wjwithagen** [comment]
@aclamk 

It sounded like you build something that would function outside of the Ceph realm as rather blunt axe to get some data.
That would have been workable to do some simple porting.

> Porting it will probably be irritating work, sorry.

Does sound like it, but then getting it to work would help iin getting some feeling for performance.

I'll see what comes of it.

**Comment by satoru-takeuchi** [comment]
@aclamk 

> BlueFS is special in the sense that it fully controls all path to data.
> While current proposal is faulty (re-reading data block that has been appended to bluefs file might be corrupted),
> in general we could fully control to always purge cache region before reading.

Keep the consistency is really difficult. For example, I've hit the following issue.

https://tracker.ceph.com/issues/51034

As a result of my investigation, I suspect that this problem is caused by the co-existence of buffered/direct I/O. The current bluefs I/O uses direct I/O even though in buffered I/O mode as follows.

https://github.com/ceph/ceph/blob/master/src/os/bluestore/BlueFS.cc#L2701

It might be fixed in the next pacific release as @ifed01 said, but it's unsure yet.

Of course, I have no right to stop merging your change. But I think that bluefs buffered I/O is not so stable and introducing more complexity may cause further hard-to-track problems.

**Comment by markhpc** [comment]
I've been thinking quite a bit about this lately. I think we need to dig in further to understand why the rocksdb block cache doesn't seem to be consistently caching data blocks on re-iteration.  That seems to be the primary culprit that's keeping us tied to buffered IO mode rather than just switching completely back to direct as we tried to do previously.  I am also wondering again if bluestore should have it's own dedicated omap cache.  It looked a bit trickier than I had hoped when I thought about this a while back, but perhaps it's worth thinking about again.
